src/webassembly.cc

```diff
@@ -316,16 +320,35 @@ void ReadDataSection(const Section& section, const IndexedNames& names,
 
   for (uint32_t i = 0; i < count; i++) {
     string_view segment = data;
-    uint8_t mode = ReadFixed<uint8_t>(&data);
+    uint32_t mode = ReadVarUInt32(&data);
     if (mode > 1) THROW("multi-memory extension isn't supported");
     if (mode == 0) { // Active segment
-      // We will need to read the init expr.
-      // For the extended const proposal, read instructions until end is reached
-      // Otherwise, just read a constexpr inst (t.const or global.get)
-      // For now, we just need to support passive segments.
-      continue;
+      // Read the initializer expression. For now we don't care about the
+      // offset, only the index (because the name section refers to the index).
+      // We don't yet support the extended-const proposal, so we only need to
+      // decode t.const or global.get
+      uint8_t opcode = ReadFixed<uint8_t>(&data);
```

Comment:
Please factor this code into a helper function. `SkipInitializerExpression()` perhaps?